gateway.go

```diff
@@ -38,75 +43,72 @@ const (
 	metricsResultInvalidContent     = "invalid_content"
 )
 
-func (s *gatewayResource) httpError(w http.ResponseWriter, status int, debugMessage string, metrics Metrics, metricsPrefix string) {
-	if s.verbose {
-		log.Println(debugMessage)
-	}
+func (s *gatewayResource) httpError(status int, debugMessage string) (ohttp.EncapsulatedResponse, HttpErr) {
+	var msg = http.StatusText(status)
 	if s.debugResponse {
-		http.Error(w, debugMessage, status)
-	} else {
-		http.Error(w, http.StatusText(status), status)
+		msg = debugMessage
 	}
-	metrics.ResponseStatus(metricsPrefix, status)
+	return ohttp.EncapsulatedResponse{}, HttpErr{status, msg}
 }
 
-func (s *gatewayResource) gatewayHandler(w http.ResponseWriter, r *http.Request) {
-	if s.verbose {
-		log.Printf("%s Handling %s\n", r.Method, r.URL.Path)
-	}
-
-	metrics := s.metricsFactory.Create(metricsEventGatewayRequest)
-
+func (s *gatewayResource) gatewayHandlerLogic(r *http.Request, metrics Metrics) (ohttp.EncapsulatedResponse, HttpErr) {
 	if r.Method != http.MethodPost {
 		metrics.Fire(metricsResultInvalidMethod)
-		s.httpError(w, http.StatusBadRequest, fmt.Sprintf("Invalid method: %s", r.Method), metrics, r.Method)
-		return
+		return s.httpError(http.StatusBadRequest, fmt.Sprintf("Invalid method: %s", r.Method))
 	}
 	if r.Header.Get("Content-Type") != ohttpRequestContentType {
 		metrics.Fire(metricsResultInvalidContentType)
-		s.httpError(w, http.StatusBadRequest, fmt.Sprintf("Invalid content type: %s", r.Header.Get("Content-Type")), metrics, r.Method)
-		return
+		return s.httpError(http.StatusBadRequest, fmt.Sprintf("Invalid content type: %s", r.Header.Get("Content-Type")))
 	}
-
 	var encapHandler EncapsulationHandler
 	var ok bool
 	if encapHandler, ok = s.encapsulationHandlers[r.URL.Path]; !ok {
-		s.httpError(w, http.StatusBadRequest, fmt.Sprintf("Unknown handler"), metrics, r.Method)
-		return
+		metrics.Fire(metricsResultInvalidContentType)
+		return s.httpError(http.StatusBadRequest, fmt.Sprintf("Unknown handler"))
 	}
-
 	defer r.Body.Close()
 	encryptedMessageBytes, err := ioutil.ReadAll(r.Body)
```

Comment:
Future improvement: consider a `LimitReader` so you only read requests up to a maximum allowed size to prevent extremely large requests from DoS'ing the gateway.

Ref: https://pkg.go.dev/io#example-LimitReader